Simplified/AxisService/NYPLAxisService.swift

```diff
@@ -75,16 +129,194 @@ class NYPLAxisService: NSObject {
+      bookVaultId: bookVaultId,
+      clientIP: deviceInfoProvider.clientIP,
+      cypher: cypher,
+      deviceID: deviceInfoProvider.deviceID,
+      isbn: isbn
+    ).generateLicenseURL()
+    
+    downloadItem(from: licenseURL, at: writeURL)
+  }
+  
+  // MARK: - Download Encryption
+  private func downloadEncryption(forBook book: NYPLBook) {
+    dispatchGroup.wait()
+    dispatchGroup.enter()
+    
+    let encryptionURL = baseURL
+      .appendingPathComponent(AxisHelper.encryptionDownloadEndpoint)
+    
+    let writeURL = self.dedicatedWriteURL
+      .appendingPathComponent(encryptionURL.lastPathComponent)
     
+    downloadItem(from: encryptionURL, at: writeURL)
+  }
+  
+  // MARK: - Download Container
+  private func downloadContainer(forBook book: NYPLBook) {
+    dispatchGroup.wait()
+    dispatchGroup.enter()
+    
+    let containerURL = baseURL
+      .appendingPathComponent(AxisHelper.containerDownloadEndpoint)
+    
+    let writeURL = self.dedicatedWriteURL
+      .appendingPathComponent(containerURL.lastPathComponent)
+    
+    downloadItem(from: containerURL, at: writeURL)
+  }
+  
+  // MARK: - Download Package
+  private func downloadPackage(forBook book: NYPLBook) {
+    dispatchGroup.wait()
+    dispatchGroup.enter()
+    guard let endpoint = packageEndpoint else {
+      dispatchGroup.leave()
+      return
+    }
+    
+    let packageURL = baseURL.appendingPathComponent(endpoint)
+    let writeURL = self.dedicatedWriteURL.appendingPathComponent(endpoint)
+    downloadItem(from: packageURL, at: writeURL)
+  }
+  
+  // MARK: - Download assets listed in package
+  private func downloadAssetsFromPackage(forBook book: NYPLBook) {
+    dispatchGroup.wait()
+    guard let packageEndpoint = packageEndpoint else {
+      // No need to log error here since an error is already logged when
+      // pacakgeEndpoint generation returns nil
+      delegate?.failDownloadWithAlert(forBook: self.book)
+      return
+    }
+    
+    let packageURL = self.dedicatedWriteURL.appendingPathComponent(packageEndpoint)
+    let data = try? Data(contentsOf: packageURL)
+    guard let axisXML = NYPLAxisXML(data: data) else {
+      NYPLErrorLogger.logError(
+        withCode: .axisDRMFulfillmentFail,
+        summary: NYPLAxisService.failedDownloadingAssetsFromPackageSummary,
+        metadata: [
+          NYPLAxisService.reason: NYPLAxisXML.NYPLXMLGenerationFailure
+      ])
+      
+      return
+    }
+    
+    let hrefs = axisXML.findRecursivelyInAttributes(AxisHelper.hrefKey)
+    
+    for href in hrefs {
+      let endpath: String
+      if let pathPrefix = packagePathPrefix {
+        endpath = "\(pathPrefix)\(href)"
+      } else {
+        endpath = href
+      }
+      
+      let linkURL = baseURL
+        .appendingPathComponent(endpath)
+      
+      let writeURL = self.dedicatedWriteURL
+        .appendingPathComponent(endpath)
+      
+      dispatchGroup.enter()
+      downloadItem(from: linkURL, at: writeURL)
```

Comment:
Here if the downloads start failing the user will see multiple alerts which is bad UX. Needs fixing

Reply:
great point... feel free to handle it in a separate ticket if you need.

Reply:
Done